### PR TITLE
Add pre attack event #4075

### PR DESCRIPTION
**Pull request by spixi**
This implements #4075

### CODE REVIEW

File: src/actions/attack.cpp [hunk 1]
```diff
@@ -1339,6 +1339,25 @@ void attack::perform()
 		return;
 	}
 
+	bc_.reset(new battle_context(units_, a_.loc_, d_.loc_, a_.weapon_, d_.weapon_));
```
Comment (jostephd) [inline src/actions/attack.cpp]: Shouldn't the `VALIDATE(a_.weapon_ < ...)` be moved up to be before this line?

Comment (spixi) [reply src/actions/attack.cpp]: I guess no, because you can change the selected attack in the `pre_attack` event.

**Comment by Pentarctagon** [comment]
@gfgtdf thoughts on this PR?

**Comment by Pentarctagon** [comment]
@CelticMinstrel any thoughts?

**Comment by CelticMinstrel** [comment]
Hmm, mostly seems fine.

There's a `refresh_bc()` call that was in the old code but is not in the new code. I have no idea whether this is a problem though.

I think we should have an example (or maybe a unit test?) where the pre_attack event changes the selected weapon.

**Pull request closed without merging**